sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,1055 @@
+	defaultOCSPResponderTimeout = 10 * time.Second
+)
+
+const (
+	// cacheExpire specifies cache data expiration time in seconds.
+	cacheExpire = float64(24 * 60 * 60)
+)
+
+const (
+	tolerableValidityRatio = 100               // buffer for certificate revocation update time
+	maxClockSkew           = 900 * time.Second // buffer for clock skew
+)
+
+type ocspCachedResponse struct {
+	time       float64
+	producedAt float64
+	thisUpdate float64
+	nextUpdate float64
+	status     ocspStatusCode
+}
+
+type Client struct {
+	// caRoot includes the CA certificates.
+	caRoot map[string]*x509.Certificate
+	// certPool includes the CA certificates.
+	certPool              *x509.CertPool
+	ocspResponseCache     *lru.TwoQueueCache
+	ocspResponseCacheLock sync.RWMutex
+	// cacheUpdated is true if the memory cache is updated
+	cacheUpdated bool
+	logFactory   func() hclog.Logger
+}
+
+type ocspStatusCode int
+
+type ocspStatus struct {
+	code ocspStatusCode
+	err  error
+}
+
+const (
+	ocspSuccess                ocspStatusCode = 0
+	ocspStatusGood             ocspStatusCode = -1
+	ocspStatusRevoked          ocspStatusCode = -2
+	ocspStatusUnknown          ocspStatusCode = -3
+	ocspStatusOthers           ocspStatusCode = -4
+	ocspFailedDecomposeRequest ocspStatusCode = -5
+	ocspInvalidValidity        ocspStatusCode = -6
+	ocspMissedCache            ocspStatusCode = -7
+	ocspCacheExpired           ocspStatusCode = -8
+)
+
+// copied from crypto/ocsp.go
+type certID struct {
+	HashAlgorithm pkix.AlgorithmIdentifier
+	NameHash      []byte
+	IssuerKeyHash []byte
+	SerialNumber  *big.Int
+}
+
+// cache key
+type certIDKey struct {
+	HashAlgorithm crypto.Hash
+	NameHash      string
+	IssuerKeyHash string
+	SerialNumber  string
+}
+
+// copied from crypto/ocsp
+var hashOIDs = map[crypto.Hash]asn1.ObjectIdentifier{
+	crypto.SHA1:   asn1.ObjectIdentifier([]int{1, 3, 14, 3, 2, 26}),
+	crypto.SHA256: asn1.ObjectIdentifier([]int{2, 16, 840, 1, 101, 3, 4, 2, 1}),
+	crypto.SHA384: asn1.ObjectIdentifier([]int{2, 16, 840, 1, 101, 3, 4, 2, 2}),
+	crypto.SHA512: asn1.ObjectIdentifier([]int{2, 16, 840, 1, 101, 3, 4, 2, 3}),
+}
+
+// copied from crypto/ocsp
+func getOIDFromHashAlgorithm(target crypto.Hash) (asn1.ObjectIdentifier, error) {
+	for hash, oid := range hashOIDs {
+		if hash == target {
+			return oid, nil
+		}
+	}
+	return nil, fmt.Errorf("no valid OID is found for the hash algorithm: %v", target)
+}
+
+func (c *Client) getHashAlgorithmFromOID(target pkix.AlgorithmIdentifier) crypto.Hash {
+	for hash, oid := range hashOIDs {
+		if oid.Equal(target.Algorithm) {
+			return hash
+		}
+	}
+	//no valid hash algorithm is found for the oid. Falling back to SHA1
+	return crypto.SHA1
+}
+
+// calcTolerableValidity returns the maximum validity buffer
+func calcTolerableValidity(thisUpdate, nextUpdate time.Time) time.Duration {
+	return durationMax(nextUpdate.Sub(thisUpdate)/tolerableValidityRatio, maxClockSkew)
```

Comment:
I've never personally seen it in a response but according to the RFC 6960 `nextUpdate` may not be set, so we should protect the code here against nils or within `isInValidityRange` below.

From https://www.rfc-editor.org/rfc/rfc6960#section-4.2.2.1
> If nextUpdate is not set, the responder is indicating that newer revocation information is available all the time.

Reply:
Addressed